File: src/task_manager/models/entities.py
"""Basic entity models for the task management system."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from uuid import UUID

from .enums import ExitCriteriaStatus, Priority, Status

DEFAULT_PROJECTS = {"Chore", "Repeatable"}


@dataclass
class Note:
    """Text annotation with content and timestamp.

    Attributes:
        content: The text content of the note
        timestamp: When the note was created
    """

    content: str
    timestamp: datetime


@dataclass
class ExitCriteria:
    """Completion condition for a task.

    Attributes:
        criteria: Description of the completion condition
        status: Current status (INCOMPLETE or COMPLETE)
        comment: Optional comment about the criteria
    """

    criteria: str
    status: ExitCriteriaStatus
    comment: Optional[str] = None


@dataclass
class Dependency:
    """Reference to another task that must be completed first.

    Attributes:
        task_id: UUID of the dependent task
        task_list_id: UUID of the task list containing the dependent task
    """

    task_id: UUID
    task_list_id: UUID


@dataclass
class ActionPlanItem:
    """Single item in an ordered action plan.

    Attributes:
        sequence: Order position in the action plan (0-indexed)
        content: Description of the action to take
    """

    sequence: int
    content: str


@dataclass
class Project:
    """Top-level organizational entity containing task lists.

    Attributes:
        id: Unique identifier for the project
        name: Human-readable name for the project
        is_default: Whether this is a default project (Chore or Repeatable)
        agent_instructions_template: Optional template for generating agent instructions
        created_at: Timestamp when the project was created
        updated_at: Timestamp when the project was last updated
    """

    id: UUID
    name: str
    is_default: bool
    created_at: datetime
    updated_at: datetime
    agent_instructions_template: Optional[str] = None

    def validate_name(self) -> bool:
        """Validate that the project name is not empty.

        Returns:
            True if the name is valid (non-empty), False otherwise
        """
        return bool(self.name and self.name.strip())

    def is_default_project(self) -> bool:
        """Check if this is a default project.

        Returns:
            True if this is a default project (Chore or Repeatable), False otherwise
        """
        return self.is_default or self.name in DEFAULT_PROJECTS


@dataclass
class TaskList:
    """Collection of related tasks within a project.

    Attributes:
        id: Unique identifier for the task list
        name: Human-readable name for the task list
        project_id: UUID of the project containing this task list
        agent_instructions_template: Optional template for generating agent instructions
        created_at: Timestamp when the task list was created
        updated_at: Timestamp when the task list was last updated
    """

    id: UUID
    name: str
    project_id: UUID
    created_at: datetime
    updated_at: datetime
    agent_instructions_template: Optional[str] = None

    def validate_name(self) -> bool:
        """Validate that the task list name is not empty.

        Returns:
            True if the name is valid (non-empty), False otherwise
        """
        return bool(self.name and self.name.strip())


@dataclass
class SearchCriteria:
    """Search and filter criteria for finding tasks.

    Attributes:
        query: Optional text to search in task titles and descriptions
        status: Optional list of status values to filter by
        priority: Optional list of priority values to filter by
        tags: Optional list of tags to filter by
        project_name: Optional project name to filter by
        limit: Maximum number of results to return (default: 50)
        offset: Number of results to skip for pagination (default: 0)
        sort_by: Sort criteria - "relevance", "created_at", "updated_at", or "priority" (default: "relevance")
    """

    query: Optional[str] = None
    status: Optional[list[Status]] = None
    priority: Optional[list[Priority]] = None
    tags: Optional[list[str]] = None
    project_name: Optional[str] = None
    limit: int = 50
    offset: int = 0
    sort_by: str = "relevance"


@dataclass
class BlockReason:
    """Information about why a task is blocked.

    Attributes:
        is_blocked: Whether the task is currently blocked
        blocking_task_ids: List of task IDs that are blocking this task
        blocking_task_titles: List of titles of tasks that are blocking this task
        message: Human-readable message explaining why the task is blocked
    """

    is_blocked: bool
    blocking_task_ids: list[UUID]
    blocking_task_titles: list[str]
    message: str


@dataclass
class DependencyAnalysis:
    """Analysis results for task dependency graphs.

    Attributes:
        critical_path: List of task IDs forming the longest chain of dependencies
        critical_path_length: Length of the critical path
        bottleneck_tasks: List of (task_id, blocked_count) tuples for tasks that block multiple other tasks
        leaf_tasks: List of task IDs that have no dependencies
        completion_progress: Percentage of completed tasks (0.0 to 100.0)
        total_tasks: Total number of tasks in the analyzed scope
        completed_tasks: Number of completed tasks in the analyzed scope
        circular_dependencies: List of cycles, where each cycle is a list of task IDs forming the cycle
    """

    critical_path: list[UUID]
    critical_path_length: int
    bottleneck_tasks: list[tuple[UUID, int]]
    leaf_tasks: list[UUID]
    completion_progress: float
    total_tasks: int
    completed_tasks: int
    circular_dependencies: list[list[UUID]]


@dataclass
class BulkOperationResult:
    """Result of a bulk operation on multiple entities.

    Attributes:
        total: Total number of operations attempted
        succeeded: Number of operations that succeeded
        failed: Number of operations that failed
        results: List of individual operation results (dicts with operation details)
        errors: List of error details for failed operations (dicts with error information)
    """

    total: int
    succeeded: int
    failed: int
    results: list[dict]
    errors: list[dict]


@dataclass
class HealthStatus:
    """Health status of the system and its components.

    Attributes:
        status: Overall system status - "healthy", "degraded", or "unhealthy"
        timestamp: When the health check was performed
        checks: Dictionary mapping check names to their results (each result is a dict with check details)
        response_time_ms: Total time taken to perform all health checks in milliseconds
    """

    status: str
    timestamp: datetime
    checks: dict[str, dict]
    response_time_ms: float


@dataclass
class Task:
    """Individual work item with title, description, status, dependencies, and exit criteria.

    Attributes:
        id: Unique identifier for the task
        task_list_id: UUID of the task list containing this task
        title: Short title describing the task
        description: Detailed description of the task
        status: Current status of the task (NOT_STARTED, IN_PROGRESS, BLOCKED, COMPLETED)
        dependencies: List of dependencies that must be completed before this task
        exit_criteria: List of conditions that must be satisfied before marking complete
        priority: Priority level of the task (CRITICAL, HIGH, MEDIUM, LOW, TRIVIAL)
        notes: List of general notes about the task
        created_at: Timestamp when the task was created
        updated_at: Timestamp when the task was last updated
        research_notes: Optional list of research notes
        action_plan: Optional ordered list of action items
        execution_notes: Optional list of execution notes
        agent_instructions_template: Optional template for generating agent instructions
        tags: List of tags for categorization and filtering
    """

    id: UUID
    task_list_id: UUID
    title: str
    description: str
    status: Status
    dependencies: list[Dependency]
    exit_criteria: list[ExitCriteria]
    priority: Priority
    notes: list[Note]
    created_at: datetime
    updated_at: datetime
    research_notes: Optional[list[Note]] = None
    action_plan: Optional[list[ActionPlanItem]] = None
    execution_notes: Optional[list[Note]] = None
    agent_instructions_template: Optional[str] = None
    tags: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        """Validate required fields after initialization.

        Raises:
            ValueError: If exit_criteria list is empty
        """
        if not self.exit_criteria:
            raise ValueError("Task must have at least one exit criteria")

    def validate_dependencies(self) -> bool:
        """Validate that dependencies list is properly formed.

        This method checks that the dependencies list exists and is a list.
        It does not check for circular dependencies (that's handled by the orchestration layer).

        Returns:
            True if dependencies is a valid list, False otherwise
        """
        return isinstance(self.dependencies, list)

    def can_mark_complete(self) -> bool:
        """Check if the task can be marked as complete.

        A task can be marked complete only if all exit criteria are marked as COMPLETE.

        Returns:
            True if all exit criteria are complete, False otherwise
        """
        if not self.exit_criteria:
            return False

        return all(
            criteria.status == ExitCriteriaStatus.COMPLETE for criteria in self.exit_criteria
        )

File: refreshui/src/core/types/entities.ts
/**
 * Core entity types for the Refresh UI task management system.
 * These types represent the domain model and are used throughout the application.
 * 
 * @module core/types/entities
 */

/**
 * Represents a project in the task management system.
 * Projects are the top-level organizational unit containing task lists.
 */
export interface Project {
  /** Unique identifier for the project */
  id: string;
  /** Display name of the project */
  name: string;
  /** Optional description of the project */
  description?: string;
  /** ISO 8601 timestamp when the project was created */
  createdAt: string;
  /** ISO 8601 timestamp when the project was last updated */
  updatedAt: string;
}

/**
 * Represents a task list within a project.
 * Task lists group related tasks together.
 */
export interface TaskList {
  /** Unique identifier for the task list */
  id: string;
  /** ID of the parent project */
  projectId: string;
  /** Display name of the task list */
  name: string;
  /** Optional description of the task list */
  description?: string;
  /** ISO 8601 timestamp when the task list was created */
  createdAt: string;
  /** ISO 8601 timestamp when the task list was last updated */
  updatedAt: string;
}

/**
 * Represents a dependency between tasks.
 * A task can depend on other tasks, forming a DAG (Directed Acyclic Graph).
 */
export interface TaskDependency {
  /** ID of the task that this task depends on */
  taskId: string;
  /** ID of the task list containing the dependent task */
  taskListId: string;
}

/**
 * Represents an exit criterion for a task.
 * Exit criteria define the conditions that must be met for a task to be considered complete.
 */
export interface ExitCriterion {
  /** Description of the criterion */
  criteria: string;
  /** Current status of the criterion */
  status: 'INCOMPLETE' | 'COMPLETE';
  /** Optional comment providing additional context */
  comment?: string;
}

/**
 * Represents a note attached to a task.
 * Notes can be general notes, research notes, or execution notes.
 */
export interface Note {
  /** Content of the note */
  content: string;
  /** ISO 8601 timestamp when the note was created */
  timestamp: string;
}

/**
 * Represents an action plan item for a task.
 * Action plan items define the steps to complete a task.
 */
export interface ActionPlanItem {
  /** Sequence number for ordering */
  sequence: number;
  /** Content/description of the action item */
  content: string;
}

/**
 * Task status values.
 */
export type TaskStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED' | 'BLOCKED';

/**
 * Task priority values.
 */
export type TaskPriority = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'TRIVIAL';

/**
 * Represents a task in the task management system.
 * Tasks are the fundamental work items that users track and complete.
 */
export interface Task {
  /** Unique identifier for the task */
  id: string;
  /** ID of the task list containing this task */
  taskListId: string;
  /** Title/name of the task */
  title: string;
  /** Optional detailed description of the task */
  description?: string;
  /** Current status of the task */
  status: TaskStatus;
  /** Priority level of the task */
  priority: TaskPriority;
  /** List of task dependencies */
  dependencies: TaskDependency[];
  /** List of exit criteria that must be met */
  exitCriteria: ExitCriterion[];
  /** General notes about the task */
  notes: Note[];
  /** Research notes gathered during task investigation */
  researchNotes: Note[];
  /** Notes recorded during task execution */
  executionNotes: Note[];
  /** Tags for categorization and filtering */
  tags: string[];
  /** Action plan items for completing the task */
  actionPlan?: ActionPlanItem[];
  /** ISO 8601 timestamp when the task was created */
  createdAt: string;
  /** ISO 8601 timestamp when the task was last updated */
  updatedAt: string;
}
